cmd/helloworld/main.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"time"
+
+	goflag "flag"
+
+	"github.com/open-cluster-management/addon-framework/pkg/version"
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilflag "k8s.io/component-base/cli/flag"
+	"k8s.io/component-base/logs"
+
+	"github.com/open-cluster-management/addon-framework/pkg/addonmanager"
+	"github.com/open-cluster-management/addon-framework/pkg/agent"
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func main() {
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	pflag.CommandLine.SetNormalizeFunc(utilflag.WordSepNormalizeFunc)
+	pflag.CommandLine.AddGoFlagSet(goflag.CommandLine)
+
+	logs.InitLogs()
+	defer logs.FlushLogs()
+
+	command := newCommand()
+	if err := command.Execute(); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+func newCommand() *cobra.Command {
+	cmd := controllercmd.
+		NewControllerCommandConfig("addon-controller", version.Get(), runController).
+		NewCommand()
+	cmd.Use = "controller"
+	cmd.Short = "Start the addon controller"
+
+	return cmd
+}
+
+func runController(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	mgr, err := addonmanager.New(controllerContext.KubeConfig)
+	if err != nil {
+		return err
+	}
+	agent := &helloWorldAgent{}
+	mgr.AddAgent(agent)
+	agentRegistration := &helloWorldAgentWithRegistration{}
+	mgr.AddAgent(agentRegistration)
+	mgr.Start(ctx)
+
+	<-ctx.Done()
+
+	return nil
+}
+
+type helloWorldAgent struct{}
+
+func (h *helloWorldAgent) Manifests(cluster *clusterv1.ManagedCluster) ([]runtime.Object, error) {
+	return []runtime.Object{
+		&corev1.ConfigMap{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ConfigMap",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "hello-world",
+				Namespace: "default",
+			},
+		},
+	}, nil
+}
+
+func (h *helloWorldAgent) GetAgentAddonOptions() *agent.AgentAddonOptions {
+	return &agent.AgentAddonOptions{
+		AddonName:             "helloworld",
+		AddonInstallNamespace: "default",
+	}
+}
+
+func (h *helloWorldAgent) GetRegistrationOption() *agent.RegistrationOption {
+	// return nil since registration is not needed by this agent
+	return nil
+}
+
+// Another agent with registration enabled.
+type helloWorldAgentWithRegistration struct{}
+
+func (h *helloWorldAgentWithRegistration) Manifests(cluster *clusterv1.ManagedCluster) ([]runtime.Object, error) {
+	return []runtime.Object{
```

Comment:
I think you may want to show how to use yaml files read in as well.  That gives people the option of describing it both ways.

Reply:
yes, I added an example to use  yaml  files.